server/src/main.rs

```diff
@@ -165,15 +188,20 @@ impl LanguageServer for Backend {
             .log_message(MessageType::INFO, "Extension settings changed")
             .await;
 
-        // When the configuration changes, any of the files in the workspace could be impacted. Therefore, we need to
-        // clear the diagnostics for all files and then re-publish them.
-        clear_diagnostics(&self.client, &self.session.configuration_sets).await;
+        // Explicit scope to ensure the session lock guard is dropped before we start compilation.
+        {
+            let mut session_guard = self.session.lock().await;
+
+            // When the configuration changes, any of the files in the workspace could be impacted. Therefore, we need to
+            // clear the diagnostics for all files and then re-publish them.
+            clear_diagnostics(&self.client, &session_guard.configuration_sets).await;
 
-        // Update the stored configuration sets from the data provided in the client notification
-        self.session.update_configurations_from_params(params).await;
+            // Update the stored configuration sets from the data provided in the client notification
+            session_guard.update_configurations_from_params(params);
+        }
 
         // Trigger a compilation and publish the diagnostics for all files
-        compile_and_publish_diagnostics(&self.client, &self.session).await;
+        self.compile_and_publish_diagnostics().await;
```

Comment:
@ReeceHumphreys

I tested it out some, and everything seems fine.

But I could totally see where the deadlock you're describing would happen.
It's why I added this 'explicit scope' here. Without it, the session would be locked once (at the top so we can update the params), and then it would try to lock a second time inside `compile_and_publish_diagnostics`. Without this scope, Rust might not drop the first lock in time - deadlock.

Sure, the code has gotten moved around and shuffled, but the basic idea of "update params, then recompile" hasn't changed.